homeassistant/components/music_assistant/services.py

```diff
@@ -0,0 +1,171 @@
+ATTR_SEARCH_ARTIST = "artist"
+ATTR_SEARCH_ALBUM = "album"
+ATTR_LIMIT = "limit"
+ATTR_LIBRARY_ONLY = "library_only"
+ATTR_FAVORITE = "favorite"
+ATTR_SEARCH = "search"
+ATTR_OFFSET = "offset"
+ATTR_ORDER_BY = "order_by"
+ATTR_ALBUM_TYPE = "album_type"
+ATTR_ALBUM_ARTISTS_ONLY = "album_artists_only"
+ATTR_CONFIG_ENTRY_ID = "config_entry_id"
+
+
+@callback
+def get_music_assistant_client(
+    hass: HomeAssistant, config_entry_id: str
+) -> MusicAssistantClient:
+    """Get the (first) Music Assistant client from the (loaded) config entries."""
+    entry: MusicAssistantConfigEntry | None
+    if not (entry := hass.config_entries.async_get_entry(config_entry_id)):
+        raise ServiceValidationError("Entry not found")
+    if entry.state is not ConfigEntryState.LOADED:
+        raise ServiceValidationError("Entry not loaded")
+    return entry.runtime_data.mass
+
+
+@callback
+def register_actions(hass: HomeAssistant) -> None:
+    """Register custom actions."""
+    register_search_action(hass)
+    register_get_library_action(hass)
+
+
+def register_search_action(hass: HomeAssistant) -> None:
+    """Register search action."""
+
+    async def handle_search(call: ServiceCall) -> ServiceResponse:
+        """Handle queue_command action."""
+        mass = get_music_assistant_client(hass, call.data[ATTR_CONFIG_ENTRY_ID])
+        search_name = call.data[ATTR_SEARCH_NAME]
+        search_artist = call.data.get(ATTR_SEARCH_ARTIST)
+        search_album = call.data.get(ATTR_SEARCH_ALBUM)
+        if search_album and search_artist:
+            search_name = f"{search_artist} - {search_album} - {search_name}"
+        elif search_album:
+            search_name = f"{search_album} - {search_name}"
+        elif search_artist:
+            search_name = f"{search_artist} - {search_name}"
+        search_results = await mass.music.search(
+            search_query=search_name,
+            media_types=call.data.get(ATTR_MEDIA_TYPE, MediaType.ALL),
+            limit=call.data[ATTR_LIMIT],
+            library_only=call.data[ATTR_LIBRARY_ONLY],
+        )
+        # return limited result to prevent it being too verbose
+        return cast(ServiceResponse, searchresults_as_compact_dict(search_results))
+
+    hass.services.async_register(
+        DOMAIN,
+        SERVICE_SEARCH,
+        handle_search,
+        schema=vol.Schema(
+            {
+                vol.Required(ATTR_CONFIG_ENTRY_ID): str,
+                vol.Required(ATTR_SEARCH_NAME): cv.string,
+                vol.Optional(ATTR_MEDIA_TYPE): vol.All(
+                    cv.ensure_list, [vol.Coerce(MediaType)]
+                ),
+                vol.Optional(ATTR_SEARCH_ARTIST): cv.string,
+                vol.Optional(ATTR_SEARCH_ALBUM): cv.string,
+                vol.Optional(ATTR_LIMIT, default=5): vol.Coerce(int),
+                vol.Optional(ATTR_LIBRARY_ONLY, default=False): cv.boolean,
+            }
+        ),
+        supports_response=SupportsResponse.ONLY,
```

Comment:
We're still registering service actions that return responses.

Reply:
oops, missed a spot. Now removed